File: DSA/Aarohi-Sharma/Day29_q2.cpp
class Solution {
public:
    int maxChunksToSorted(vector<int>& arr) {
        int n = arr.size();
        vector<int> left_max(n, 0);
        left_max[0] = arr[0];
        for(int i = 1; i < n; i++)
        {
            left_max[i] = max(left_max[i - 1], arr[i]);
        }
        vector<int> right_min(n, 0);
        right_min[n - 1] = arr[n - 1];
        for(int i = n - 2; i >= 0; i--)
        {
            right_min[i] = min(right_min[i + 1], arr[i]);
        }
        int count = 0;
        for(int i = 0; i < n - 1; i++)
        {
            if(left_max[i] <= right_min[i + 1])
            {
                count++;
            }
        }
        count++;
        return count;
    }
};

File: DSA/Ishita-Nigam/Day29/Day29_Q2.java
public class Day29_Q2 {
    public int maxChunksToSorted(int[] arr) {
        int[] rmin = new int[arr.length + 1];
        rmin[arr.length] = Integer.MAX_VALUE;
        
        for (int i = arr.length - 1; i >= 0; i--) {
            rmin[i] = Math.min(rmin[i + 1], arr[i]);
        }
        
        int lmax = Integer.MIN_VALUE;
        int count = 0;
        
        for (int i = 0; i < arr.length; i++) {
            lmax = Math.max(lmax, arr[i]);
            
            if (lmax <= rmin[i + 1]) {
                count++;
            }
        }
        return count;
    }

    public static void main(String[] args) {
        Day29_Q2 solution = new Day29_Q2();
        int[] arr = {2, 1, 3, 4, 4};
        int result = solution.maxChunksToSorted(arr);
        System.out.println("Maximum number of chunks: " + result);
    }
}
